python/packages/declarative/agent_framework_declarative/_workflows/_graph/_base.py

```diff
@@ -0,0 +1,685 @@
+"""
+
+import logging
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, cast
+
+from agent_framework._workflows import (
+    Executor,
+    SharedState,
+    WorkflowContext,
+)
+
+logger = logging.getLogger(__name__)
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+# Key used in SharedState to store declarative workflow variables
+DECLARATIVE_STATE_KEY = "_declarative_workflow_state"
+
+
+class DeclarativeWorkflowState:
+    """Manages workflow variables stored in SharedState.
+
+    This class provides the same interface as the interpreter-based WorkflowState
+    but stores all data in SharedState for checkpointing support.
+
+    The state is organized into namespaces:
+    - workflow.inputs: Initial inputs (read-only)
+    - workflow.outputs: Values to return from workflow
+    - turn: Variables persisting within the workflow turn
+    - agent: Results from most recent agent invocation
+    - conversation: Conversation history
+    """
+
+    def __init__(self, shared_state: SharedState):
+        """Initialize with a SharedState instance.
+
+        Args:
+            shared_state: The workflow's shared state for persistence
+        """
+        self._shared_state = shared_state
+
+    async def initialize(self, inputs: "Mapping[str, Any] | None" = None) -> None:
+        """Initialize the declarative state with inputs.
+
+        Args:
+            inputs: Initial workflow inputs (become workflow.inputs.*)
+        """
+        state_data: dict[str, Any] = {
+            "inputs": dict(inputs) if inputs else {},
+            "outputs": {},
+            "turn": {},
+            "agent": {},
+            "conversation": {"messages": [], "history": []},
+            "custom": {},
+        }
+        await self._shared_state.set(DECLARATIVE_STATE_KEY, state_data)
+
+    async def get_state_data(self) -> dict[str, Any]:
+        """Get the full state data dict from shared state."""
+        try:
+            result: dict[str, Any] = await self._shared_state.get(DECLARATIVE_STATE_KEY)
+            return result
+        except KeyError:
+            # Initialize if not present
+            await self.initialize()
+            return cast(dict[str, Any], await self._shared_state.get(DECLARATIVE_STATE_KEY))
+
+    async def set_state_data(self, data: dict[str, Any]) -> None:
+        """Set the full state data dict in shared state."""
+        await self._shared_state.set(DECLARATIVE_STATE_KEY, data)
+
+    async def get(self, path: str, default: Any = None) -> Any:
+        """Get a value from the state using a dot-notated path.
+
+        Args:
+            path: Dot-notated path like 'turn.results' or 'workflow.inputs.query'
+            default: Default value if path doesn't exist
+
+        Returns:
+            The value at the path, or default if not found
+        """
+        # Map .NET style namespaces to Python style
+        if path.startswith("Local."):
+            path = "turn." + path[6:]
```

Comment:
nit: perhaps if we do 
```suggestion
            path = "turn." + path[len("Local."):]
```
, it'd be easier to manage